[dev-2.0] Fix optional and rest parameters in TypeScript class method declarations

SoundOfScooting opened the pull request:
<!--
  Thank you for contributing! Please use this pull request (PR) template.


 In the description field of this PR, include "resolves #XXXX" tagging the issue you are fixing. If this PR addresses the issue but doesn't completely resolve it (ie the issue should remain open after your PR is merged), write "addresses #XXXX".-->
Addresses #7862 (fixes overloads but not missing fields or other syntax errors)

 Changes:
<!-- Add here what changes were made in this pull request and if possible provide links showcasing the changes. -->
* Adds type declaration filepaths to `package.json` so they can be imported.
* Fixes class method declarations not emitting optional and rest parameters.

 Screenshots of the change:
<!-- If applicable, add screenshots depicting the changes. -->

#### PR Checklist
<!--
  To check any option, replace the "[ ]" with a "[x]". Be sure to check out how it looks in the Preview tab! Feel free to remove any portion of the template that is not relevant for your issue.
-->

- [x] `npm run lint` passes
- [ ] [Inline reference] is included / updated
- [ ] [Unit tests] are included / updated

[Inline reference]: https://p5js.org/contribute/contributing_to_the_p5js_reference/
[Unit tests]: https://github.com/processing/p5.js/tree/main/contributor_docs#unit-tests

limzykenneth commented on package.json:
```diff
@@ -68,8 +68,14 @@
   "exports": {
-    ".": "./dist/app.js",
-    "./core": "./dist/core/main.js",
+    ".": {
```
I'm guessing for the other exports we don't have a setup for their types just yet that would make sense? We don't need it now as the full ESM modular use is not at a stage where I want to push it publicly yet, but we can start thinking about how it would work.

I'm fully in favor of including `"type": "module"` in package.json and shift any remaining code we have that still uses cjs to ESM, I think dependencies wise it should all support ESM without issue now.

SoundOfScooting commented:
Working on fixing more type resolution and syntax errors, plus a few reference tag corrections.
I've not been organized and will still have to divide all the changes into commits.

One question I have is what I should do about `@private` classes whose types are referenced in public methods - some examples in `p5.d.ts` are `Renderer`, `RendererGL`, `Texture`, `FramebufferTexture`, `Quat`, `Cubic`, `ImageInfo[]` (which is wrong and I've corrected it to `ImageInfos`, although still private).
Since the class is not emitted it will fail to resolve the type (even after the resolution fixes), and the class's public methods are incorrectly emitted as regular functions.

For some of these I'm also wondering if they should even be private in the first place. For instance, the reference page for [`image()`](https://beta.p5js.org/reference/p5/image/) links to missing pages for private classes. Alternatively we could implement a new tag (e.g. `@internal`) which would hide a class from the reference but not the declaration files?

SoundOfScooting commented:
Some of the private types like `ImageInfos` were only being used in local functions that were erroneously detected to be methods of `p5`, so I've currently marked them `@private` and removed `@method`. Not sure what to do about the abstract classes (I think anything private doesn't even get seen by `helper.mjs`?) so I'm leaving them as-is.

To fix type resolution, I have tried two potential avenues so far:
1) Emit references to `p5.Type` as `Type` instead (`p5` ends up refering to a class and not the module). This will fail in `global.d.ts` since it's declarations aren't inside the module, so this fix should not be applied to that file, but I didn't succeed in conditionally applying it and switched to fix 2 (see below for a fix that would work though).
2) Inspired by `@types/p5`, in `p5.d.ts` change `export default p5` to `export = p5`. I haven't really tested this for ESM imports (I don't think they'd even work to begin with since you're supposed to import from the export subpaths directly?), but it works correctly in non-module `<script>` blocks unlike 1 (which I've changed my game to use so I could drop its dependency on a bundler).
  Previously, you would have a globally declared namespace `p5` that contains the `p5` class as a default export, so only this would work at runtime:
  ```ts
  let pInst: p5.default = new (p5 as unknown as typeof p5.default)(p => {...})
  ```

Another change I currently have applied is changing `global.d.ts` to not redeclare everything from scratch but instead copy them from the `p5` import. This seems like a lot less work and bypasses the issue with fix 1. For example:
```ts
/** imagine docs here */
function describe(text: string, display?: p5.FALLBACK | p5.LABEL): void;
// -->
/** imagine docs here */
let describe: p5["describe"]; // syntax for fix 2, different for 1
```
This works as long as the `p5` class has no accessors (which I see no way to declare inside `declare global` but could still be in `interface Window`).

There's also the issue of both `p5.d.ts` and `core/main.d.ts` declaring `class p5` (the former being complete and latter incomplete). We could just hardcode `main.d.ts` to not declare anything, but I say we again look to `@types/p5` here and have `p5.d.ts` declare an empty `class p5` with `interface p5 extends some_merged_interface {}` to remove this duplication as well as align `p5` method declarations with their corresponding JS files.
